### PR TITLE
[Tbbas-1183] builder pattern improvement (add constructor with builder argument)

### PR DESCRIPTION
**Pull request by denise-opendaq**
## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [ ] Pull request title reflects its content
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

# Description:

### CODE REVIEW

File: core/coreobjects/include/coreobjects/property_impl.h [hunk 1]
```diff
@@ -114,6 +115,33 @@ class PropertyImpl : public ImplementationOf<IProperty, ISerializable, IProperty
         checkErrorInfo(validateDuringConstruction());
     }
 
+    PropertyImpl(IPropertyBuilder* propertyBuilder)
```
Comment (JakaMohorkoDS) [inline core/coreobjects/include/coreobjects/property_impl.h:118]: These constructors should now replace the ones using dictionaries imo - those were really only created for use within the `build` method. 

Those dictionary constructors should be removed on all builder-pattern objects.

File: core/coreobjects/include/coreobjects/property_impl.h [hunk 2]
```diff
@@ -114,6 +115,33 @@ class PropertyImpl : public ImplementationOf<IProperty, ISerializable, IProperty
...
+        this->visible = propertyBuilderPtr.getVisible();
+        this->readOnly = propertyBuilderPtr.getReadOnly();
+        this->selectionValues = propertyBuilderPtr.getSelectionValues();
+        this->suggestedValues = propertyBuilderPtr.getSuggestedValues();
+        this->refProp = propertyBuilderPtr.getReferencedProperty();
+        this->coercer = propertyBuilderPtr.getCoercer();
+        this->validator = propertyBuilderPtr.getValidator();
+        this->callableInfo = propertyBuilderPtr.getCallableInfo();
+        this->onValueWrite = EventEmitter(propertyBuilderPtr.getOnPropertyValueWrite());
+        this->onValueRead = EventEmitter(propertyBuilderPtr.getOnPropertyValueRead());
```
Comment (JakaMohorkoDS) [inline core/coreobjects/include/coreobjects/property_impl.h:137]: Is the EventEmitter wrapper required here?

**Review by JakaMohorkoDS — CHANGES_REQUESTED** [review]
Looks good to me. Only the dictionary constructors/factories should be cleaned up still. Would also ask @Unsttopabull to check.

**Comment by JakaMohorkoDS** [comment]
Also, changes to the changelog document are missing.

File: core/coreobjects/include/coreobjects/property_impl.h (hunk 2 above)
Comment (Unsttopabull) [reply core/coreobjects/include/coreobjects/property_impl.h:137]: It shouldn't be necessary, otherwise the `operator=` needs to be fixed.

**Review by Unsttopabull — CHANGES_REQUESTED** [review]

File: core/opendaq/signal/include/opendaq/data_descriptor_impl.h [hunk 3]
```diff
@@ -86,6 +85,8 @@ class DataDescriptorImpl : public GenericStructImpl<IDataDescriptor, IStruct, IS
     void initCalcs();
     std::unique_ptr<ScalingCalc> scalingCalc;
     std::unique_ptr<DataRuleCalc> dataRuleCalc;
+
+    static DictPtr<IString, IBaseObject> packBuilder(IDataDescriptorBuilder* dataDescriptorBuilder);
```
Comment (Unsttopabull) [inline core/opendaq/signal/include/opendaq/data_descriptor_impl.h:89]: Styleguide: Static member methods are UpperCase

**Review by Unsttopabull — APPROVED** [review]
Looks good, it just needs some coding-style changes

**Review by JakaMohorkoDS — CHANGES_REQUESTED** [review]
The implementation looks really good to me. Currently only some tests are missing. 

- The new factories are inherently tested by the existing tests that call `build`, but it wouldn't hurt to try them out by passing a builder to the factory in a test to isolate that use case.
- Similarly, the getters are tested in the same manner, but an explicit test testing that a set value is also returned by the corresponding getter.

Since the functionality is already tested in some manner, I'm okay with merging the branch. It's up to you if you want to add more tests.

**Review by JakaMohorkoDS — APPROVED** [review]

**Pull request merged**